### DISCUSSION TITLE
Uncaught Error: The submit event is dispatched by form#a instead of form#b

### DISCUSSION THREAD

**Discussion by marilari88** [edmundhung/conform#590] [Q&A]
### Describe the bug and the expected behavior

When attempting to submit a form within a radix-ui dialog, I encountered an error message in the console: "Uncaught Error: The submit event is dispatched by form#b instead of form#a". The form#a resides within the dialog.

To illustrate this issue, I've provided a reproducible example here: [link](https://stackblitz.com/~/github.com/marilari88/conform-with-dialog).

It appears that despite only submitting form#a, the submit event handler of form#b is triggered.

### Conform version

v1.1.0

### Steps to Reproduce the Bug or Issue

1. Open [https://stackblitz.com/~/github.com/marilari88/conform-with-dialog](https://stackblitz.com/~/github.com/marilari88/conform-with-dialog) or clone locally if you prefere [this](github.com/marilari88/conform-with-dialog)
2. Press 'add player'
3. Fill the name and goal field
4. Submit

### What browsers are you seeing the problem on?

Chrome

### Screenshots or Videos

![conform error](https://github.com/edmundhung/conform/assets/32909388/dc2df843-15f5-41cb-a8b7-e8679cf50852)

**Comment by edmundhung** [comment]
Interesting! This might be a bug on React? It can be resolved by moving the dialog out of the team form. Although the dialog will renders the player form outside of the team form, but the react submit event still bubbles up through the virtual dom, which triggers the `onSubmit` handler on the team form and trigger the error you had.

**Comment by houmark** [comment]
Nested forms is not allowed in HTML, so that appears to be the base reason for the error.

I read in the docs that you should be able to use a [FormProvider/FormContext](https://conform.guide/api/react/FormProvider#tips) in Conform to control which form you submit, but I have not tested this myself and wonder why @edmundhung did not suggest that as a solution, so maybe I am misunderstanding something here. If that's the case, then please disregard, but just wanted to mention a possible solution if you are not able to move your dialog outside the parent form.

**Comment by edmundhung** [comment]
The forms are never nested in the HTML markup because of the dialog rendering the player form outside through a portal. It's only the synthetic event from react bubbles up on the virtual dom which cause this issue - An submit event dispatched on the player form get bubbles up to team form and trigger the onSubmit handler on both form.

FormContext only helps with accessing the form state and won't help here.

**Comment by penx** [comment]
I think you can use stopPropagation to solve this?

```ts
const [form, fields] = useForm({
  id: useId(),
  onSubmit: (e, context) => {
    e.stopPropagation();
  },
});
```

edit: not ideal though - [caveats](https://css-tricks.com/dangers-stopping-event-propagation)

**Comment by penx** [comment]
I've also encountered this issue today. Our use case is we have a large form, some buttons within it that open a modal.

The modal contains a small form, representing a part of the parent form data. The user can fill this in or cancel out of the modal. This is a different form as, if the user cancels out, it should not affect the data of the parent form.

As explained above, this is valid html as the child form is within the React component tree but not the DOM of the parent form.

We make use of `useInputControl`'s `change` to update the parent form when the modal form is submitted.

We want to use conform for both forms to keep validation and data structure consistent.

I have't yet found a solution other than using `e.stopPropagation`.

Other things I've tried that don't seem to work but may be worth further research:

- setting form={form.id} on the submit button
- removing the `type=submit` from the button and calling `conformFormProps.onSubmit` directly

**Comment by edmundhung** [comment]
Nested form tag is a bad practice in general so it is probably better to avoid it even your Dialog might render it outside. Conform has full support of form attribute and is utilizing the formId generated to make sure all your inputs and intent button works no matter they are enclosing within the form tag. You should be able to keep it working like this:

```tsx
export function Component() {
  const [form, fields] = useForm({
    // ...
  });

  return (
    <>
       <Form {...getFormProps(form)} />
       <input {...getInputProps(fields.example)} />
       <button form={form.id}>Submit</button>
    </>
  );
}
```

I am converting this to an discussion since the issue wasn't a bug in Conform.

**Reply by marilari88** [reply]
hi @edmundhung your approch is a nice escamotage for the issue I reported, but I found a problem with the onChange event of the form element. It seems like it is not triggered by changing the inputs outside the form tags

**Comment by luciengeorge** [comment]
I've been having a similar issue. In my case I didn't have nested forms but I was getting this warning in my console.

<img width="590" alt="Screenshot 2024-10-23 at 4 43 48 pm" src="https://github.com/user-attachments/assets/d4a85d0e-4670-4635-ba6b-392692ead13d">

If I look at the form in my DOM, I could see that the id of the form was `:RoslH2:`. And then when submitting the form I would get the following error

<img width="609" alt="Screenshot 2024-10-23 at 4 46 00 pm" src="https://github.com/user-attachments/assets/85caca38-506b-4cb6-b2e8-126cb1c1c530">

I managed to fix by hardcoding an `id` to the form. using `useId()` didn't work 🤔 
```ts
const [form, fields] = useForm({
  id: 'nav-form',
 // ...
})
```

I don't know what was causing the issue in the first place, but this fixed it for me. I am using `v1.2.2` of conform.
Hope this helps anyone having the same issue but would really like to understand the underlying issue as well

**Comment by bmthd** [comment]
I really wanted to use nested forms, so I investigated ways to make them work.
We succeeded in getting it to work by making the following modifications.
The key is to stop the bubble-up not onSubmit of the useForm, but onSubmit of its return value.

```tsx
export const Form: FC<
  Omit<ComponentProps<"form">, keyof ReturnType<typeof getFormProps>> & {
    form: FormMetadata<any>;
  }
> = ({ form, ...props }) => {
  const { onSubmit, ...formProps } = getFormProps(form);
  const handleSubmit = useCallback<FormEventHandler<HTMLFormElement>>(
    (event) => {
      event.stopPropagation();
      onSubmit(event);
    },
    [onSubmit],
  );
  return (
    <FormProvider context={form.context}>
      <form {...props} {...formProps} onSubmit={handleSubmit} />
    </FormProvider>
  );
```
